File: communicator/communicator_push/src/main/java/ru/tensor/sbis/communicator/di/quickreply/QuickReplyComponent.java
package ru.tensor.sbis.communicator.di.quickreply;

import androidx.annotation.Nullable;

import dagger.BindsInstance;
import dagger.Component;
import ru.tensor.sbis.communicator.common.di.CommunicatorCommonComponent;
import ru.tensor.sbis.communicator.quickreply.QuickReplyManager;
import ru.tensor.sbis.communicator.quickreply.QuickReplyModel;
import ru.tensor.sbis.communicator.ui.quickreply.QuickReplyContract;
import ru.tensor.sbis.pushnotification.center.PushCenter;
import ru.tensor.sbis.user_activity_track.service.UserActivityService;

/**
 * Created by aa.mironychev on 08.08.17.
 */
@QuickReplyScope
@Component(
        modules = { QuickReplyModule.class },
        dependencies = {
                CommunicatorCommonComponent.class
        }
)
public interface QuickReplyComponent {

    QuickReplyContract.Presenter getPresenter();

    QuickReplyManager getManager();

    PushCenter getPushCenter();

    @Component.Builder
    interface Builder {
        @BindsInstance
        Builder quickReplyModel(@Nullable QuickReplyModel model);

        @BindsInstance
        Builder userActivityService(@Nullable UserActivityService userActivityService);

        Builder communicatorCommonComponent(CommunicatorCommonComponent communicatorCommonComponent);

        QuickReplyComponent build();
    }
}

File: communicator/features/communicator_crm_conversation/src/main/java/ru/tensor/sbis/communicator/crm/conversation/presentation/ui/quick_reply/di/QuickReplyComponent.kt
package ru.tensor.sbis.communicator.crm.conversation.presentation.ui.quick_reply.di

import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.LifecycleCoroutineScope
import androidx.lifecycle.ViewModelStoreOwner
import dagger.BindsInstance
import dagger.Component
import dagger.assisted.AssistedFactory
import ru.tensor.sbis.common.di.CommonSingletonComponent
import ru.tensor.sbis.communicator.crm.conversation.presentation.ui.quick_reply.ui.QuickReplyController
import ru.tensor.sbis.communicator.crm.conversation.presentation.ui.quick_reply.ui.QuickReplyView
import ru.tensor.sbis.communicator.declaration.crm.model.QuickReplyParams

/**
 * @author dv.baranov
 */

/** @SelfDocumented */
@QuickReplyScope
@Component(
    dependencies = [CommonSingletonComponent::class],
    modules = [QuickReplyModule::class],
)
internal interface QuickReplyComponent {

    fun injector(): Injector

    val viewFactory: (View) -> QuickReplyView

    @Component.Factory
    interface Factory {
        fun create(
            commonSingletonComponent: CommonSingletonComponent,
            @BindsInstance viewModelStoreOwner: ViewModelStoreOwner,
            @BindsInstance quickReplyParams: QuickReplyParams,
            @BindsInstance scope: LifecycleCoroutineScope,
        ): QuickReplyComponent
    }

    @AssistedFactory
    interface Injector {
        fun inject(
            fragment: Fragment,
            viewFactory: (View) -> QuickReplyView,
        ): QuickReplyController
    }
}
